test/common/stats/lazyinit_stats_test.cc

```diff
@@ -0,0 +1,126 @@
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "source/common/common/thread.h"
+#include "source/common/stats/lazy_init.h"
+#include "source/common/stats/thread_local_store.h"
+
+#include "test/test_common/utility.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Stats {
+namespace Test {
+
+#define AWESOME_STATS(COUNTER, GAUGE, HISTOGRAM, TEXT_READOUT, STATNAME) COUNTER(foo)
+
+MAKE_STAT_NAMES_STRUCT(AwesomeStatNames, AWESOME_STATS);
+MAKE_STATS_STRUCT(AwesomeStats, AwesomeStatNames, AWESOME_STATS);
+
+class LazyInitStatsTest : public testing::Test {
+public:
+  SymbolTableImpl symbol_table_;
+  AllocatorImpl allocator_{symbol_table_};
+  ThreadLocalStoreImpl store_{allocator_};
+  AwesomeStatNames stats_names_{symbol_table_};
+};
+
+using LazyAwesomeStats = LazyCompatibleStats<AwesomeStats>;
+
+TEST_F(LazyInitStatsTest, StatsGoneWithScope) {
+  {
+    ScopeSharedPtr scope = store_.createScope("bluh");
+    LazyAwesomeStats x = LazyAwesomeStats::create(scope, stats_names_, true);
+    LazyAwesomeStats y = LazyAwesomeStats::create(scope, stats_names_, true);
+    EXPECT_EQ(TestUtility::findGauge(store_, "bluh.AwesomeStats.initialized")->value(), 0);
+    x->foo_.inc();
+    EXPECT_EQ(TestUtility::findGauge(store_, "bluh.AwesomeStats.initialized")->value(), 1);
+    EXPECT_EQ(x->foo_.value(), 1);
+    EXPECT_EQ(y->foo_.value(), 1);
+    EXPECT_EQ(TestUtility::findGauge(store_, "bluh.AwesomeStats.initialized")->value(), 2);
+  }
+  // Deleted as scope deleted.
+  EXPECT_EQ(TestUtility::findCounter(store_, "bluh.foo"), nullptr);
+  {
+    ScopeSharedPtr scope = store_.createScope("bluh");
+    LazyAwesomeStats x = LazyAwesomeStats::create(scope, stats_names_, true);
+    EXPECT_EQ(TestUtility::findGauge(store_, "bluh.AwesomeStats.initialized")->value(), 0);
+    EXPECT_EQ(x->foo_.value(), 0);
+    // Initialized now.
+    EXPECT_EQ(TestUtility::findGauge(store_, "bluh.AwesomeStats.initialized")->value(), 1);
+  }
+}
+
+TEST_F(LazyInitStatsTest, StatsMutlipleInstancesDynamicallyDestructed) {
+  {
+    ScopeSharedPtr scope_1 = store_.createScope("bluh");
+    auto x =
+        std::make_unique<LazyAwesomeStats>(LazyAwesomeStats::create(scope_1, stats_names_, true));
+    auto y =
+        std::make_unique<LazyAwesomeStats>(LazyAwesomeStats::create(scope_1, stats_names_, true));
+    EXPECT_EQ(TestUtility::findGauge(store_, "bluh.AwesomeStats.initialized")->value(), 0);
+    (*x)->foo_.inc();
+    EXPECT_EQ(TestUtility::findGauge(store_, "bluh.AwesomeStats.initialized")->value(), 1);
+    EXPECT_EQ((*x)->foo_.value(), 1);
+    EXPECT_EQ((*y)->foo_.value(), 1);
+    // x,y instantiated.
+    EXPECT_EQ(TestUtility::findGauge(store_, "bluh.AwesomeStats.initialized")->value(), 2);
+    // Only x instantiated.
+    y.reset();
```

Comment:
have you tried to repro the race per our chat?

you could do that with thread_synchronizer though that might be annoying since this is a template class.

You could also try to just run the case I outlined a bunch of times, with the bug still in the code, to repro it, then re-fix the bug and prove it's fixed.